File: tks/RestAdapters/src/main/kotlin/rest/api/Producer.java
package rest.api;

import rest.api.adapters.RentRepositoryAdapter;
import rest.api.adapters.ResourceRepositoryAdapter;
import rest.api.adapters.UserRepositoryAdapter;
import data.AbstractResourceEntity;
import data.RentEntity;
import data.UserEntity;
import rest.api.mappers.RentMapper;
import rest.api.mappers.ResourceMapper;
import rest.api.mappers.UserMapper;
import ports.primary.ResourceRentCommandPort;
import ports.secondary.ResourcePersistencePort;
import ports.secondary.ResourceSearchPort;
import ports.secondary.UserPersistencePort;
import ports.secondary.UserSearchPort;
import repositories.RepositoryBase;
import rest.api.application.services.RentService;
import rest.api.application.services.ResourcesService;
import rest.api.application.services.UserService;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.inject.Produces;
import javax.inject.Singleton;
import java.util.ArrayList;

@ApplicationScoped
public class Producer {

    @Produces
    public UserService produceUserService(UserPersistencePort userPersistencePort, UserSearchPort userSearchPort){
        return new UserService(userPersistencePort, userSearchPort);
    }

    @Produces
    public ResourcesService produceResourceService(ResourcePersistencePort resourcePersistencePort, ResourceSearchPort resourceSearchPort, RentRepositoryAdapter rentRepositoryAdapter){
        return new ResourcesService(resourcePersistencePort, resourceSearchPort, rentRepositoryAdapter);
    }

    @Produces
    public UserRepositoryAdapter produceUserSavePort(RepositoryBase<UserEntity> repository, UserMapper mapper){
        return new UserRepositoryAdapter(repository, mapper);
    }

    @Produces
    public ResourceRepositoryAdapter produceResourceRepository(RepositoryBase<AbstractResourceEntity> repository, ResourceMapper mapper){
        return new ResourceRepositoryAdapter(repository, mapper);
    }

    @Produces
    public ResourceRentCommandPort produceResourceRentCommandPort(RentRepositoryAdapter adapter, UserRepositoryAdapter userRepositoryAdapter, ResourceRepositoryAdapter resourceRepositoryAdapter){
        return new RentService(adapter, adapter, userRepositoryAdapter, resourceRepositoryAdapter);
    }

    @Produces
    public RentRepositoryAdapter produceRentRepositoryAdapter(RepositoryBase<UserEntity> userRepository, RepositoryBase<AbstractResourceEntity> resourceRepository, RepositoryBase<RentEntity> rentRepository, RentMapper rentMapper){
        return new RentRepositoryAdapter(rentRepository,resourceRepository, userRepository, rentMapper);
    }

    @Produces
    @Singleton
    public RepositoryBase<UserEntity> produceUserRepository(){
        var list = new ArrayList<UserEntity>();
        return new RepositoryBase<>(list);
    }

    @Produces
    @Singleton
    public RepositoryBase<AbstractResourceEntity> produceResourceRepository(){
        var list = new ArrayList<AbstractResourceEntity>();
        return new RepositoryBase<>(list);
    }

    @Produces
    @Singleton
    public RepositoryBase<RentEntity> produceRentRepository(){
        var list = new ArrayList<RentEntity>();
        return new RepositoryBase<>(list);
    }

    @Produces
    public UserMapper produceUserMapper(){
        return UserMapper.Companion.getINSTANCE();
    }

    @Produces
    public ResourceMapper produceResourceMapper(){
        return ResourceMapper.Companion.getINSTANCE();
    }

    @Produces
    public RentMapper produceRentMapper(){
        return RentMapper.Companion.getINSTANCE();
    }
}

File: tks/RestAdapters/src/main/kotlin/rest/api/Producer.kt
package rest.api

import application.services.RentService
import application.services.ResourcesService
import application.services.UserService
import ports.secondary.combined.IResourceRepositoryAdapter
import ports.secondary.combined.IUserRepositoryAdapter
import repository.adapters.RentRepositoryAdapter
import repository.adapters.ResourceRepositoryAdapter
import repository.adapters.UserRepositoryAdapter
import repository.data.AbstractResourceEntity
import repository.data.BookEntity
import repository.data.RentEntity
import repository.data.UserEntity
import repository.mappers.RentMapper
import repository.mappers.ResourceMapper
import repository.mappers.UserMapper
import repository.repositories.RepositoryBase
import java.util.*
import javax.enterprise.context.ApplicationScoped
import javax.enterprise.inject.Produces
import javax.inject.Singleton

@ApplicationScoped
class Producer {


    private var resources: MutableList<AbstractResourceEntity>
    private var users: MutableList<UserEntity>
    private var rents: MutableList<RentEntity>

    init {
        val guid = UUID.fromString("7b4399fe-5f73-40fe-90a4-1163f3dfc221")
        val user = UserEntity(UUID.randomUUID(), "mszewc@edu.pl", "ADMIN", "password", true)
        val book = BookEntity(UUID.randomUUID(), "EEEE-254", "Diuna", "Frank Herbert")
        val rent = RentEntity(UUID.randomUUID(), guid, Date(), null, user, book)
        rents = mutableListOf(rent)
        users = mutableListOf(user)
        resources = mutableListOf(book)
    }

    @Produces
    fun produceUserService(adapter: IUserRepositoryAdapter): UserService = UserService(adapter, adapter)

    @Produces
    fun produceResourceService(
        resourceAdapter: IResourceRepositoryAdapter,
        rentRepositoryAdapter: RentRepositoryAdapter
    ): ResourcesService = ResourcesService(resourceAdapter, resourceAdapter, rentRepositoryAdapter)

    @Produces
    fun produceRentService(
        userAdapter: IUserRepositoryAdapter,
        resourceAdapter: IResourceRepositoryAdapter,
        rentAdapter: RentRepositoryAdapter
    ): RentService = RentService(rentAdapter, rentAdapter, userAdapter, resourceAdapter)

    @Produces
    fun produceUserRepositoryAdapter(
        repository: RepositoryBase<UserEntity>,
        mapper: UserMapper
    ): UserRepositoryAdapter = UserRepositoryAdapter(repository, mapper)

    @Produces
    fun produceResourceRepositoryAdapter(
        repository: RepositoryBase<AbstractResourceEntity>,
        mapper: ResourceMapper
    ): ResourceRepositoryAdapter = ResourceRepositoryAdapter(repository, mapper)

    @Produces
    fun produceRentRepositoryAdapter(
        userRepository: RepositoryBase<UserEntity>,
        resourceRepository: RepositoryBase<AbstractResourceEntity>,
        rentRepository: RepositoryBase<RentEntity>,
        rentMapper: RentMapper
    ): RentRepositoryAdapter = RentRepositoryAdapter(rentRepository, resourceRepository, userRepository, rentMapper)

    @Produces
    @Singleton
    fun produceUserRepository(): RepositoryBase<UserEntity> = RepositoryBase(users)

    @Produces
    @Singleton
    fun produceResourceRepository(): RepositoryBase<AbstractResourceEntity> = RepositoryBase(resources)

    @Produces
    @Singleton
    fun produceRentRepository(): RepositoryBase<RentEntity> = RepositoryBase(rents)

    @Produces
    fun produceUserMapper(): UserMapper = UserMapper.INSTANCE

    @Produces
    fun produceResourceMapper(): ResourceMapper = ResourceMapper.INSTANCE

    @Produces
    fun produceRentMapper(): RentMapper = RentMapper.INSTANCE
}